### PR TITLE
Fix v1 2 3 upgrade tests

### PR DESCRIPTION
**Pull request by dylant-da**
Fixes from https://github.com/digital-asset/daml/pull/18977 to make the tests on v1-2-3 upgrades check what they're supposed to. Previously, we'd run two asynchronous assertions, but only the last one was being actually used. This fixes those tests.

### CODE REVIEW

File: sdk/daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/UpgradesSpec.scala [hunk 1]
```diff
@@ -22,6 +22,7 @@ import com.daml.lf.archive.DarReader
 
 import scala.util.{Success, Failure}
 import com.daml.lf.validation.Upgrading
+import org.scalatest.Inspectors.{forEvery}
```
Comment (carlpulley-da) [inline sdk/daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/UpgradesSpec.scala:25]: Nitpick: do we need the curly parens here?

File: sdk/daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/UpgradesSpec.scala [hunk 2]
```diff
@@ -332,59 +368,91 @@ trait LongTests { this: UpgradesSpec =>
...
-        "test-common/upgrades-SuccessUpgradingV2ThenV3-v1.dar",
-        "test-common/upgrades-SuccessUpgradingV2ThenV3-v3.dar",
-        assertPackageUpgradeCheck(None),
-      )
+      for {
+        v1Upload <- uploadPackage("test-common/upgrades-SuccessUpgradingV2ThenV3-v1.dar")
+        v2Upload <- uploadPackage("test-common/upgrades-SuccessUpgradingV2ThenV3-v2.dar")
+        v3Upload <- uploadPackage("test-common/upgrades-SuccessUpgradingV2ThenV3-v3.dar")
+      } yield {
+        val cantonLog = Source.fromFile(s"$cantonTmpDir/canton.log")
```
Comment (carlpulley-da) [inline sdk/daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/UpgradesSpec.scala:376]: In Java, we would write something like:
```
try (cantonLog = Source.fromFile(s"$cantonTmpDir/canton.log")) {
 // etc.
}
```
and have it manage the file resource for us. Sadly, this isn't directly possible in Scala.

However, I believe we can write:
```
try (Source cantonLog = Source.fromFile(s"$cantonTmpDir/canton.log")) {
  // etc.
}
```
or:
```
import scala.util.Using
Using(Source.fromFile(s"$cantonTmpDir/canton.log")) { cantonLog =>
  // etc.
}
```

File: sdk/daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/UpgradesSpec.scala [hunk 3]
```diff
@@ -332,59 +368,91 @@ trait LongTests { this: UpgradesSpec =>
...
-        "test-common/upgrades-SuccessUpgradingV3ThenV2-v2.dar",
-        assertPackageUpgradeCheck(None),
-      )
+      for {
+        v1Upload <- uploadPackage("test-common/upgrades-SuccessUpgradingV3ThenV2-v1.dar")
+        v3Upload <- uploadPackage("test-common/upgrades-SuccessUpgradingV3ThenV2-v3.dar")
+        v2Upload <- uploadPackage("test-common/upgrades-SuccessUpgradingV3ThenV2-v2.dar")
+      } yield {
+        val cantonLog = Source.fromFile(s"$cantonTmpDir/canton.log")
+        try {
```
Comment (carlpulley-da) [inline sdk/daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/UpgradesSpec.scala:398]: ditto

File: sdk/daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/UpgradesSpec.scala [hunk 4]
```diff
@@ -332,59 +368,91 @@ trait LongTests { this: UpgradesSpec =>
...
-          Some("The upgraded template T has added new fields, but those fields are not Optional.")
-        ),
-      )
+      for {
+        v1Upload <- uploadPackage("test-common/upgrades-FailsWhenUpgradingV2ThenV3-v1.dar")
+        v2Upload <- uploadPackage("test-common/upgrades-FailsWhenUpgradingV2ThenV3-v2.dar")
+        v3Upload <- uploadPackage("test-common/upgrades-FailsWhenUpgradingV2ThenV3-v3.dar")
+      } yield {
+        val cantonLog = Source.fromFile(s"$cantonTmpDir/canton.log")
+        try {
```
Comment (carlpulley-da) [inline sdk/daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/UpgradesSpec.scala:419]: ditto

File: sdk/daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/UpgradesSpec.scala [hunk 5]
```diff
@@ -332,59 +368,91 @@ trait LongTests { this: UpgradesSpec =>
...
-          Some("The upgraded template T has added new fields, but those fields are not Optional.")
-        ),
-      )
+      for {
+        v1Upload <- uploadPackage("test-common/upgrades-FailsWhenUpgradingV3ThenV2-v1.dar")
+        v3Upload <- uploadPackage("test-common/upgrades-FailsWhenUpgradingV3ThenV2-v3.dar")
+        v2Upload <- uploadPackage("test-common/upgrades-FailsWhenUpgradingV3ThenV2-v2.dar")
+      } yield {
+        val cantonLog = Source.fromFile(s"$cantonTmpDir/canton.log")
+        try {
```
Comment (carlpulley-da) [inline sdk/daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/UpgradesSpec.scala:442]: ditto

File: sdk/daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/UpgradesSpec.scala [hunk 6]
```diff
@@ -604,24 +691,29 @@ abstract class UpgradesSpec(val suffix: String)
...
+    val (testPackageV2Id, uploadV2Result) = v2
     uploadV1Result should be(empty)
     cantonLogSrc should include(
       s"Ignoring upload of package $testPackageV2Id as it has been previously uploaded"
     )
     uploadV2Result should be(empty)
   }
 
+  def assertDontCheckUpload(
+      @annotation.unused v1: (PackageId, Option[Throwable]),
```
Comment (carlpulley-da) [inline sdk/daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/UpgradesSpec.scala:707]: Oh nice, I didn't know about that annotation

**Review by carlpulley-da — APPROVED** [review]
Thanks for this, modulo a little dry up, LGTM!

File: sdk/daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/UpgradesSpec.scala (hunk 2 above)
Comment (dylant-da) [reply sdk/daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/UpgradesSpec.scala:376]: I've opted to read the file entirely via `Using` before trying any assertions over it, like

```
...
        v3Upload <- uploadPackage("test-common/upgrades-SuccessUpgradingV3ThenV2-v3.dar")
        v2Upload <- uploadPackage("test-common/upgrades-SuccessUpgradingV3ThenV2-v2.dar")
        rawCantonLog <- Future.fromTry(Using(Source.fromFile(s"$cantonTmpDir/canton.log"))(_.mkString))
      } yield {
        forEvery(
...
```

**Pull request merged**